File: algorithms/implementation/circular-array-rotation/circular-array-rotation-javascript/js/circularArrayRotation.js
// # Você tem um array de números inteiros e um número fixo de rotações. 
// # A tarefa é rotacionar o array à direita por um número especificado de vezes.

// # Um array de números inteiros a.
// # Um inteiro k, que representa o número de rotações à direita a serem executadas.
// # Um array de consultas queries, onde cada elemento é um índice que deve ser consultado após as rotações.

// s

//     // Em JavaScript, você pode usar arrays para armazenar os resultados.
//     let resultado = [];

//     // Para criar um array de números inteiros em JavaScript, basta atribuir os valores diretamente.
//     // Por exemplo:
//     // let arrayOriginal = [1, 2, 3, 4, 5];

//     // No JavaScript moderno, você pode usar a função map para realizar operações em cada elemento de um array.
//     // Aqui está um exemplo para simular o comportamento do ArrayList em Java:
//     let arrayOriginal = a.map(Number);

//     let mudancas = k % arrayOriginal.length;

//     for (let x of consultas) {
//         resultado.push(arrayOriginal[(x - mudancas + arrayOriginal.length) % arrayOriginal.length]);
//     }

//     // A variável 'resultado' agora contém os resultados desejados.
//     // console.log(resultado);

//     return resultado;

// }


// Define uma função chamada circularArrayRotation que recebe 
// um array 'a', um inteiro 'k' e uma lista de consultas 'consultas'.

export function circularArrayRotation(a, k, consultas) {

    // Realiza a rotação 'k' vezes.
    for (let index = 1; index <= k; index++) {

        // Remove o último elemento do array 'a'.
        const numero = a.pop();

        // Adiciona o elemento removido no início do array 'a'.
        a.unshift(numero);
        
    }

    // Mapeia a lista de consultas para os elementos 
    // correspondentes no array 'a' rotacionado e retorna o resultado.
    return consultas.map(q => a[q]);
}

File: algorithms/implementation/circular-array-rotation/circular-array-rotation-python/tests/circularArrayRotation.py
# Você tem um array de números inteiros e um número fixo de rotações. 
# A tarefa é rotacionar o array à direita por um número especificado de vezes.

# Um array de números inteiros a.
# Um inteiro k, que representa o número de rotações à direita a serem executadas.
# Um array de consultas queries, onde cada elemento é um índice que deve ser consultado após as rotações.


# # A função recebe três parâmetros: um array a ser rotacionado, o número de rotações (k) e uma lista de consultas.

# def circularArrayRotation(a, rotacao, consultas):
    
#     # Inicializa uma lista vazia para armazenar os resultados das consultas.
#     resultado = []
    
#     # Calcula o número real de rotações necessário usando o operador módulo (%).
#     # Isso evita rotações desnecessárias e mantém o resultado dentro dos limites do array.
#     mudancas = rotacao % len(a)
    
#     # Itera sobre a lista de consultas.
#     for x in consultas:
        
#         # Para cada consulta, adiciona o elemento do array após as rotações ao resultado.
#         resultado.append(a[x - mudancas])
    
#     # Retorna a lista de resultados das consultas após as rotações.
#     return resultado


# Função que realiza a rotação circular de uma lista e retorna os elementos correspondentes às consultas

def circularArrayRotation(a, rotacao, consultas):
    
    # Realiza a rotação 'rotacao' vezes
    for _ in range(rotacao):
        
        # Remove o último elemento da lista e o insere no início
        numero = a.pop()
        
        a.insert(0, numero)
        
    # Lista de compreensão para obter os elementos correspondentes às consultas na lista rotacionada
    resultado = [a[q] for q in consultas]
    
    # Retorna a lista de resultados
    return resultado